dsp/modules/premai.py

```diff
@@ -0,0 +1,151 @@
+import logging
+from typing import Optional
+
+import backoff
+import premai.errors
+
+from dsp.modules.lm import LM
+
+try:
+    import premai
+except ImportError as err:
+    raise ImportError(
+        "Not loading Mistral AI because it is not installed. Install it with `pip install premai`.",
+    ) from err
+
+
+# Set up logging
+logger = logging.getLogger(__name__)
+logging.basicConfig(level=logging.INFO)
+
+
+class ChatPremAPIError(Exception):
+    """Error with the `PremAI` API."""
+
+
+ERROR = ChatPremAPIError
+
+
+def backoff_hdlr(details) -> None:
+    """Handler for the backoff package.
+
+    See more at: https://pypi.org/project/backoff/
+    """
+    logger.info(
+        "Backing off {wait:0.1f} seconds after {tries} tries calling function {target} with kwargs {kwargs}".format(
+            **details,
+        ),
+    )
+
+
+def giveup_hdlr(details) -> bool:
+    """Wrapper function that decides when to give up on retry."""
+    if "rate limits" in details.message:
+        return False
+    return True
+
+
+class PremAI(LM):
+    """Wrapper around Prem AI's API."""
+
+    def __init__(
+        self,
+        model: str,
+        project_id: int,
+        api_key: str,
+        base_url: Optional[str] = None,
+        session_id: Optional[int] = None,
+        **kwargs,
+    ) -> None:
+        """Parameters
+
+        model: str
+            The name of model name
+        project_id: int
+            "The project ID in which the experiments or deployments are carried out. can find all your projects here: https://app.premai.io/projects/"
+        api_key: str
+            Prem AI API key, to connect with the API.
+        session_id: int
+            The ID of the session to use. It helps to track the chat history.
+        **kwargs: dict
+            Additional arguments to pass to the API provider
+        """
+        super().__init__(model)
+        self.kwargs = kwargs if kwargs == {} else self.kwargs
+
+        self.project_id = project_id
+        self.session_id = session_id
+
+        if base_url is not None:
+            self.client = premai.Prem(api_key=api_key, base_url=base_url)
+        else:
+            self.client = premai.Prem(api_key=api_key)
+        self.provider = "premai"
+
```

Comment:
missing the line `self.history: list[dict[str, Any]] = []`

Reply:
Thanks for pointing this out, but I wonder, `self.history` is also been initialized in the abstract base class, so why we also define it here?

Reply:
defining the history here gives some flexibility with the typing (that is not done in the LM class). This behavior is maintained in most of the other LM classes ([gpt3.py](https://github.com/stanfordnlp/dspy/blob/0e595a70cd9d14b0db25252320aad11eae08fd72/dsp/modules/gpt3.py#L94), [Ollama](https://github.com/stanfordnlp/dspy/blob/main/dsp/modules/ollama.py), etc.).

Feel free to test the PremAI behavior with lm.inspect_history(n=1) on your end (where lm is defined as your dspy.PremAI model) as well.